internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+
+	return s, nil
+}
+
+// Run until the given context is canceled. This method blocks, so should be
+// called in a goroutine.
+func (s *Store) Run(ctx context.Context) { s.pub.Run(ctx) }
+
+// Read a resource using its ID.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) Read(id *pbresource.ID) (*pbresource.Resource, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	val, err := tx.First(tableNameResources, indexNameID, id)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return nil, storage.ErrNotFound
+	}
+
+	res := val.(*pbresource.Resource)
+
+	// Observe the Uid if it was given.
+	if id.Uid != "" && res.Id.Uid != id.Uid {
+		return nil, storage.ErrNotFound
+	}
+
+	// Let the caller know they need to upgrade/downgrade the schema version.
+	if id.Type.GroupVersion != res.Id.Type.GroupVersion {
+		return nil, storage.GroupVersionMismatchError{
+			RequestedType: id.Type,
+			Stored:        res,
+		}
+	}
+
+	return res, nil
+}
+
+// WriteCAS performs an atomic Check-And-Set (CAS) write of a resource.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) WriteCAS(res *pbresource.Resource, vsn string) error {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	existing, err := tx.First(tableNameResources, indexNameID, res.Id)
+	if err != nil {
+		return err
+	}
+
+	if existing == nil && vsn != "" {
+		return storage.ErrConflict
+	}
+
+	if existing != nil {
+		existingRes := existing.(*pbresource.Resource)
+
+		// Ensure CAS semantics.
+		if existingRes.Version != vsn {
+			return storage.ErrConflict
+		}
+
+		// Uid is immutable.
+		if existingRes.Id.Uid != res.Id.Uid {
+			return storage.ErrConflict
+		}
+	}
+
+	if err := tx.Insert(tableNameResources, res); err != nil {
+		return err
+	}
+
+	idx, err := incrementEventIndex(tx)
+	if err != nil {
+		return nil
+	}
+	tx.Commit()
+
+	s.publishEvent(idx, &pbresource.WatchEvent{
+		Operation: pbresource.WatchEvent_OPERATION_UPSERT,
+		Resource:  res,
+	})
+
+	return nil
+}
+
+// DeleteCAS performs an atomic Check-And-Set (CAS) deletion of a resource.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) DeleteCAS(id *pbresource.ID, vsn string) error {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	existing, err := tx.First(tableNameResources, indexNameID, id)
+	if err != nil {
+		return err
+	}
+
+	// Deleting an already deleted resource is a no-op.
+	if existing == nil {
+		return nil
+	}
+
+	res := existing.(*pbresource.Resource)
+
+	// Deleting a resource using a previous Uid is a no-op.
+	if id.Uid != res.Id.Uid {
+		return nil
+	}
+
+	// Ensure CAS semantics.
+	if vsn != res.Version {
+		return storage.ErrConflict
+	}
+
+	if err := tx.Delete(tableNameResources, id); err != nil {
+		return err
+	}
+
+	idx, err := incrementEventIndex(tx)
+	if err != nil {
+		return nil
+	}
+	tx.Commit()
+
+	s.publishEvent(idx, &pbresource.WatchEvent{
```

Comment:
❗ (probably a blocker)

With event publishing after the `tx.Commit` isn't it theoretically possible to see events out-of-order. 

If using this Store in a single-threaded fashion like we currently do with Raft + MemDB then concurrent writes wont be happening anyways so its *extremely* unlikely it happens as it would require the go routine to be paused for long enough between the `tx.Commit` and the `publishEvent` calls happen for another write + publish to occur. 

As this is the in-mem store and it appears like there are no guarantees of consistency or write serialization, it feels like this is more likely to happen. That in turn could be a source of flakey tests.

Do you think it would be too difficult to integrate the event publishing bits in here with memdbs changeset tracking to ensure that events are published in the same order that the txns occur.

Reply:
Good catch. Just moving the call to `publishEvent` before the transaction is committed should do the trick (which is what the changeset tracking does anyway).

https://github.com/hashicorp/consul/blob/d5677e5680050a9b22403c160f5edde59613a4fc/agent/consul/state/memdb.go#L164-L170

Reply:
We discussed this a bunch offline and having the publish after memdb txn commit should be fine iff we also ensure that no more than one write operation against the store is happening.

Basically we need to add a `sync.Mutex` to the `Store` and lock it for the duration of the `WriteCAS` and `DeleteCAS` methods. One optimization to that could be to lock it immediately before the `txn.Commit` calls and relinquish the lock once the events have been published. MemDB will already prevent two write txns from progressing at the same time and and not gaining the lock until the end will allow the memdb txn preparation concurrently with the previous writers event publishing. However that seems like it may be premature optimization so just locking the whole method also seems fine.